File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length-1;
        
        return recur(nums, target, start, end);
}
 
    public int recur(int[] nums, int target, int left, int right) {
        if(left > right)
            return -1;

        int mid = left + (right - left)/2;

        if(nums[mid] == target)
            return mid;
        else if(nums[mid] > target) {
            right = mid-1;
            return recur(nums, target, left, right);
        }else {
            left = mid+1;
            return recur(nums, target, left, right);
        }
    }

}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        """

        start, end - initialize 

        x - 1
        y - 2


        length of the window - top occurance value in the map <= k 

        else 

        shrink the left side of the array
        delete the values from the map

        calc max value and return max value
        """

        res = 0
        left = 0
        right = 0
        map = {}


        while right < len(s):
            map[s[right]] = map.get(s[right], 0) + 1

            # get the max_value 
            max_occurance_value = self.calcMaxValue(map)
            if (right-left+1) - max_occurance_value <= k:
                res = max(res, (right-left+1))
                right +=1 
            else:
                # delete from map 
                map[s[left]] -= 1
                if map[s[left]] == 0:
                    del map[s[left]]
                left += 1
                right +=1                 


        return res



        

    def calcMaxValue(self, map):
        max_value = 0
        for key,value in map.items():
            max_value = max(max_value, value)
        
        return max_value
